[FLINK-7650] [flip6] Port JobCancellationHandler to new REST endpoint

tillrohrmann opened the pull request:
## What is the purpose of the change

Let the `JobCancellationHandler` implement the `LegacyRestHandler` interface. Moreover,
this commit adds the `DELETE` method to `HttpMethodWrapper` and the
`RestServerEndpoint#registerHandler` method. This PR also adds the `DispatcherGateway#cancelJob` and `JobMasterGateway#cancel` RPC calls.

## Brief change log

- Add `DispatcherGateway#cancelJob` and `JobMasterGateway#cancel` RPC calls
- Let `JobCancellationHandler` implement `LegacyRestHandler`
- Introduce `JobMessageParameters`
- Introduce `JobCancellationHeaders`
- Add `JobCancellationHandler` to `DispatcherRestEndpoint`

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

zentol commented on flink-runtime-web/src/test/resources/log4j-test.properties:
```diff
@@ -16,7 +16,7 @@
 #
 
-log4j.rootLogger=OFF, console
+log4j.rootLogger=INFO, console
```
revert

tillrohrmann replied:
Good catch. Sorry that I missed that. Will revert it.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterGateway.java:
```diff
@@ -55,6 +55,14 @@
 public interface JobMasterGateway extends CheckpointCoordinatorGateway, FencedRpcGateway<JobMasterId> {
 
 	/**
+	 * Cancel the currently executed job.
```
it would be more consistent with the javadoc of other method to say "Cancels"

tillrohrmann replied:
Hmm, not sure if that is true tbh. We actually have quite a wild mix of how we write JavaDocs in `JobMasterGateway`. I will change all of them starting with a verb in the 3rd person.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterGateway.java:
```diff
@@ -55,6 +55,14 @@
+	 * Cancel the currently executed job.
+	 *
+	 * @param timeout of this operation
+	 * @return Future acknowledge if the cancellation was successful
```
what is returned if the cancellation was not successful?

tillrohrmann replied:
at the moment it cannot fail

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobPathParameter.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Path parameter identifying jobs.
+ *
+ * <p>The type of this parameter is {@link JobID}.
```
I think we can remove this.

tillrohrmann replied:
true

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/JobCancellationHandler.java:
```diff
@@ -70,4 +82,25 @@ public JobCancellationHandler(Executor executor, Time timeout) {
+
+					throw new CompletionException(
+						new RestHandlerException(
+							"Job cancellation failed: " + error.getMessage(),
```
Is the only reason this can fail due to the job not existing?

tillrohrmann replied:
It could also be due to a `TimeoutException`. I'll add more detailed exception handling to generate the proper `HttpResponseStatus` values.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobCancellationHeaders.java:
```diff
@@ -0,0 +1,68 @@
+
+	@Override
+	public HttpMethodWrapper getHttpMethod() {
+		return HttpMethodWrapper.DELETE;
```
I still think this should not be a DELETE since we can still access information about that job under the URL that we just issues the delete call to, even after the delete has finished.

tillrohrmann replied:
Hmm, I see your point, but I don't have good feelings to put everything behind a patch method with an endless list of query parameters. This seems somewhat wrong to me. Maybe the problem is rather that we don't distinguish between running and archived jobs on the API level. Many things don't make sense to be called on an archived job, e.g. taking a savepoint. Thus, it might make more sense to distinguish that via `/running_jobs/` and `/archived_jobs/`.

zentol replied:
what "endless list of query parameters" are you referring to?

tillrohrmann replied:
For example: `rescale`, `restart`, `pause`. Others to come in the future. Agreed that if this list contains two more methods, it wouldn't change much.

bowenli86 commented on flink-runtime/src/main/java/org/apache/flink/runtime/messages/JobNotFoundException.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Exception which is returned if a job could not be found.
+ */
+public class JobNotFoundException extends FlinkException {
```
I prefer `FlinkJobNotFoundException`

tillrohrmann replied:
Have no strong feelings about it and can rename it.

bowenli86 commented on flink-runtime/src/main/java/org/apache/flink/runtime/messages/JobNotFoundException.java:
```diff
@@ -0,0 +1,34 @@
+	private static final long serialVersionUID = -7803390762010615384L;
+
+	public JobNotFoundException(JobID jobId) {
+		super("Could not find job (" + jobId + ").");
```
I feel "Could not find Flink job ..." is better

tillrohrmann replied:
will change it

bowenli86 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobCancellationHeaders.java:
```diff
@@ -0,0 +1,68 @@
+
+	@Override
+	public String getTargetRestEndpointURL() {
+		return "/jobs/:jobid";
```
to avoid typo and easier code identification, please make this a public/private static final String

tillrohrmann replied:
Good idea.

bowenli86 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobPathParameter.java:
```diff
@@ -0,0 +1,45 @@
+ *
+ * <p>The type of this parameter is {@link JobID}.
+ */
+public class JobPathParameter extends MessagePathParameter<JobID> {
```
I was actually going to create the same class for https://issues.apache.org/jira/browse/FLINK-7694, as described in that ticket :)

I prefer `JobIDPathParameter`

tillrohrmann replied:
Alright. Will rename it.

tillrohrmann commented:
Thanks for the review @bowenli86 and @zentol. I've addressed your PR comments modulo the `DELETE` method. I somehow still think that `DELETE` is the right verb and we maybe should change how we represent running and archived jobs differently. What do you think?

zentol commented:
I see a few problems with separating running&completed jobs.

* The monitoring pattern for a job becomes quite weird. While a job is running you ask "/jobs/running/:jobid", until you get a 404 and then check "/jobs/completed/:jobid".
* Submitting a job is a POST to "/jobs", which returns "/jobs/running/:jobid" which may return a 404 if the job fails straight away.
* A unified URL allows us to actually provide good error messages, i.e. a 405 Method not Allowed when trying to take a savepoint for finished jobs. For separate URLs this would return a 404.
* We have to handle the case of a savepoint/stop request arriving for the finished job anyway, if the job finishes between accepting the request and handling it.

Given that it only affects taking savepoints and stopping a job i don't think it's really necessary to change things. Note that if we were to do that the job termination should probably be a POST instead.

tillrohrmann commented:
I guess some of the problems like taking a savepoint could still be served under `/jobs/:jobid/savepoints/:savepointid` because this resource will be removed upon first request anyway.

I'm not sure how well providing better error messages will work in the end, because right now, we don't distinguish between completed and running jobs on the handler side. What it currently does is trigger `takeSavepoint` on the `Dispatcher`. The `Dispatcher` will fail with a `FlinkJobNotFoundException` if the job is no longer executed/non-existent. In order to see whether there was a job being executed, the handler then had to access some other information

Monitoring a job is indeed a bit tricky with splitting the URLs. Theoretically, this is already the case, since we only store a limited number of `ArchivedExecutionGraphs` on the `MemoryArchivist`. Thus, if you have many completing jobs, it might be the case that you don't see it when querying `/jobs/:jobid`.

Nevertheless, the advantages of keeping it like it is are evident. I'll change the method to `PATCH`. Let's see how far we get with that.

tillrohrmann commented:
Thanks for the review @zentol. I've rebased it and wait for Travis.

bowenli86 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobMessageParameters.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.Collections;
+
+/**
+ * Parameters for job related REST handlers.
```
"Parameters for job `ID` related REST handlers"?

bowenli86 commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobMessageParameters.java:
```diff
@@ -0,0 +1,42 @@
+ *
+ * <p>A job related REST handler always requires a {@link JobIDPathParameter}.
+ */
+public class JobMessageParameters extends MessageParameters {
```
sorry i missed this file name yesterday. How about `JobIDMessageParameters`? since there're no job query params in this class

zentol replied:
This class is meant to be a general parameters class for whatever parameters are generally required for job-related queries. While it currently only contains the job ID i think it makes sense to leave it as is.

bowenli86 replied:
ok, got you

zentol commented:
+1 from my side.
